### PR TITLE
Random video input device + tidy-ups for Video::Input code

### PR DESCRIPTION
**Pull request by alexdewar**
This is a tiny PR that mostly just adds a ``Video::Input`` device that just returns random white noise with a specified image size, for debugging etc. (I was having trouble using my crappy old webcam with OpenCV and adding this was easier!)

I've also included some random misc tidy-ups for ``Video::Input`` classes.

### CODE REVIEW

File: video/randominput.h [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+private:
+    cv::Size m_Size;
+    std::string m_CameraName;
+    std::random_device m_RandomDevice;
+    std::mt19937 m_Generator;
+    std::uniform_int_distribution<uchar> m_Distribution;
+
+    void fillRandom(uchar *start, uchar *end)
```
Comment (neworderofjamie) [inline video/randominput.h]: Perhaps unnecessary C++ary but:
``std::generate(start, end, [this](){ return m_Distribution(m_Generator); });``

File: video/randominput.h [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+// OpenCV
+#include <opencv2/opencv.hpp>
+
+// Standard C++ includes
+#include <random>
+
+namespace BoBRobotics {
+namespace Video {
+//! A Video::Input which generates white noise, for testing purposes
+class RandomInput : public Input {
```
Comment (neworderofjamie) [inline video/randominput.h]: As it's header-only anyway, you **could** template on the type of generator.

File: video/opengl.h [hunk 3]
```diff
@@ -46,16 +46,14 @@ class OpenGL : public Input
 
     virtual bool readFrame(cv::Mat &outFrame) override
     {
-        // If outFrame is not of the expected size or format, recreate it
-        if (outFrame.cols != m_ReadWidth || outFrame.rows != m_ReadHeight || outFrame.type() != CV_8UC3) {
-            outFrame.create(m_ReadHeight, m_ReadWidth, CV_8UC3);
-        }
+        // Make sure frame is of right size and type
+        outFrame.create(m_ReadWidth, m_ReadHeight, CV_8UC3);
```
Comment (neworderofjamie) [inline video/opengl.h]: I was initially paranoid/suspicious of this, but it's clear from the docs that this is a good idea as create has exactly this behaviour :smile:

File: video/randominput.h (hunk 1 above)
Comment (alexdewar) [reply video/randominput.h]: Hey, finding an opportunity to use STL algorithms is what C++ is about! This is def nicer

File: video/randominput.h (hunk 2 above)
Comment (alexdewar) [reply video/randominput.h]: That's true. I did think about this, but the only thing is that I can't seem to find any other generators that would work here (i.e. return ``uchar``s in the right range). I guess you could provide a custom generator though?

File: video/opengl.h (hunk 3 above)
Comment (alexdewar) [reply video/opengl.h]: Yeah, I only just discovered this :-)

File: video/randominput.h (hunk 2 above)
Comment (neworderofjamie) [reply video/randominput.h]: ``std::uniform_int_distribution<uchar>`` is totally the only distribution that makes sense but different RNGs could plausible be useful (basically cos MT is slow)

File: video/randominput.h (hunk 1 above)
Comment (neworderofjamie) [reply video/randominput.h]: That is totally my understanding of Herb Sutter's intentions :smile:

**Pull request merged**